### ISSUE TITLE
Support for multiple statements?

### ISSUE THREAD

**Issue by arackaf** [planetscale/database-js#85]
Apologies if this is already supported, but would it be possible to get the ability to run multiple statements in one request? Something like 

```js
await conn.execute("INSERT INTO foo(x) VALUES(12); SELECT LAST_INSERT_ID() as id;")
```

Obviously you can achieve that with separate queries, but that would require blocking until the insert is done, and then blocking while you grab the inserted id.

**Comment by mattrobenolt** [comment]
The issue currently is needing support for this on the server. I plan on adding some multi statement API but haven't gotten to it yet. It'd be more like a pipeline where the server would just execute them sequentially to avoid round trips.

**Comment by arackaf** [comment]
Oof - look forward to seeing that! Thanks!

**Comment by mattrobenolt** [comment]
Correct, I'm not sure why this is closed either, I'm guessing in error.

I assume once we get the server support, it'll be easy to add in here.

**Comment by arackaf** [comment]
Gotcha - just glad I'm not crazy. I don't have access to re-open, so feel free to, unless you have this work tracked elsewhere.

Also just curious how high of a prio this is these days. Seems like it's be a *huge* perf win, letting you do batch operations (ie, insert ids into a temp table, then run two queries against it) without server roundtrips.

**Comment by arackaf** [comment]
Yeah I guess that can be tricky. If I send a batch of 5 queries, the first 4 succeed, the 5th throws an error ....

**Comment by mattrobenolt** [comment]
fwiw I've had an internal issue open for the server side implementation for this feature. So it's definitely on my mind.

**Comment by juriadams** [comment]
In case anyone else has to manually migrate data into Planetscale or just has any other use case where they want to insert multiple records at the same time, here's a non-edge solution using `mysql2`:

#### `utils/database.ts`
```ts
import * as mysql from "mysql2/promise";

export const createDirectConnection = async (): Promise<
    mysql.Connection & { close: () => Promise<void> }
> => {
    const connection = await mysql.createConnection(process.env.DATABASE_URL);

    return Object.assign(connection, {
        close: connection.end,
    });
};
```

#### `index.ts`
```ts
import { createDirectConnection } from "./utils/database";

const inserts = [
    [
        1,
        "your",
        "values",
        "here",
    ],
    [
        2,
        "your",
        "values",
        "here",
    ],
];

await database.query(
    "INSERT INTO table (`id`, `column1`, `column2`, `column3`) VALUES ?",
    [inserts]
);
```

I've tried the above solution with batches of 1000 records each, working flawlessly. Though, inserts of this size will take exponentially longer — keep that in mind.

Also, inside the `database.query` function, `inserts` is wrapped by another pair of `[]`. This is very important to insert multiple records simultaneously in `mysql2`! (Meaning, the second parameter is actually `[ [ [ 1, ... ], [2, ... ] ] ]`)

I'm debating on hosting my own proxy on Cloud Run, which handles batch inserts exclusively, as this is a crucial feature for any database.

In my project, I have to insert data into multiple tables simultaneously. This goes up to 25 records, taking up to 1500ms **per transaction**, which is everything but acceptable for an API.

**Comment by mattrobenolt** [comment]
Reopening because this isn't implemented and I still would like to see this through.

The API to support this still needs to be shipped first before implementation can be done in here, but the client implementation should be rather simple.

**Comment by kneelsdev** [comment]
I would like to see this as well.

**Comment by vince-winkintel** [comment]
Same here, I'd like to be able to batch inserts and updates.

**Comment by focux** [comment]
Is this still planned?

**Comment by anszom** [comment]
+1

this would be a great performance boost in my use case. Something like:
```
UPDATE something SET counter = counter + 1 WHERE blah;
SELECT * FROM something WHERE blah;
```
should not require two round-trips.

This becomes even more interesting with transactions. With some careful design, a full transaction could be sent within one request, massively reducing the total duration of the transaction from the server's perspective.

I understand that handling failures may be a tricky problem, but even a naive solution could be useful. Essentially equivalent to a native mysql client sending multiple single-query requests without waiting for a response in-between.

**Comment by sbbu** [comment]
I would like to see this too...
